command/loader.js

```diff
@@ -0,0 +1,339 @@
+      programId: UPGRADEABLE_BPF_LOADER_PROGRAM_ID,
+      data: encodeInstruction({ SetAuthority: {} }),
+    });
+  }
+
+  static async getDeployBufferTransaction(
+    connection,
+    payer,
+    data,
+    program,
+    bufferAccount,
+    authority,
+  ) {
+    const programSpace = 36; // UpgradeableLoaderState::program_len()
+    const programBalanceNeeded =
+      await connection.getMinimumBalanceForRentExemption(programSpace);
+    const [programDataKey, _nonce] = PublicKey.findProgramAddressSync(
+      [program.toBuffer()],
+      UPGRADEABLE_BPF_LOADER_PROGRAM_ID,
+    );
+
+    const deployTransaction = new Transaction()
+      .add(
+        SystemProgram.createAccount({
+          fromPubkey: payer,
+          newAccountPubkey: program,
+          lamports: programBalanceNeeded,
+          space: programSpace,
+          programId: UPGRADEABLE_BPF_LOADER_PROGRAM_ID,
+        }),
+      )
+      .add(
+        new TransactionInstruction({
+          keys: [
+            { pubkey: payer, isSigner: true, isWritable: true },
+            { pubkey: programDataKey, isSigner: false, isWritable: true },
+            { pubkey: program, isSigner: false, isWritable: true },
+            {
+              pubkey: bufferAccount,
+              isSigner: false,
+              isWritable: true,
+            },
+
+            { pubkey: SYSVAR_RENT_PUBKEY, isSigner: false, isWritable: false },
+            { pubkey: SYSVAR_CLOCK_PUBKEY, isSigner: false, isWritable: false },
+            {
+              pubkey: SystemProgram.programId,
+              isSigner: false,
+              isWritable: false,
+            },
+            { pubkey: authority, isSigner: true, isWritable: false },
+          ],
+          programId: UPGRADEABLE_BPF_LOADER_PROGRAM_ID,
+          data: encodeInstruction({
+            DeployWithMaxDataLen: { max_data_len: data.length * 3 },
+          }),
+        }),
+      );
+    return deployTransaction;
+  }
+}
+
+async function initBuffer(connection, payer, authority, data, bufferAccount) {
+  console.log("buffer account", bufferAccount.publicKey.toBase58());
+  console.log("authority account", authority.publicKey.toBase58());
+  // UpgradeableLoaderState::buffer_len(program_len) = 37 + program_len
+  const bufferSpace = 37 + data.length;
+  const balanceNeeded =
+    await connection.getMinimumBalanceForRentExemption(bufferSpace);
+
+  const initTransaction = new Transaction()
+    .add(
+      SystemProgram.createAccount({
+        fromPubkey: payer.publicKey,
+        newAccountPubkey: bufferAccount.publicKey,
+        lamports: balanceNeeded,
+        space: bufferSpace,
+        programId: UPGRADEABLE_BPF_LOADER_PROGRAM_ID,
+      }),
+    )
+    .add(
+      new TransactionInstruction({
+        keys: [
+          {
+            pubkey: bufferAccount.publicKey,
+            isSigner: false,
+            isWritable: true,
+          },
+          { pubkey: authority.publicKey, isSigner: false, isWritable: false },
```

Comment:
Without marking the authority as a signer (and signing the transaction), InitializeBuffer will fail when the loader checks authority.is_signer.


```suggestion
          { pubkey: authority.publicKey, isSigner: true, isWritable: false },
```

Reply:
no the authority is not signer for initBuffer.